### ISSUE TITLE
Align APIs of Store and Dataset

### ISSUE THREAD

**Issue by Natureshadow** [oxigraph/oxigraph#713]
**Is your feature request related to a problem? Please describe.**

I am trying to write a wrapper around Oxigraph and OxRDF that can both store quads in memory using OxRDF types and in RocksDB.

**Describe the solution you'd like**

It would be helpful if `Store` and `Dataset` could share the same API, using a common trait. Especially, it would be helpful if `Store` could return `GraphView`s from its data.

**Describe alternatives you've considered**

An alternative would be to re-implement all sorts of methods in my wrapper API, which would defeat the usefulness of using a library like Oxigraph at some point.

**Comment by Tpt** [comment]
Thank you so much for building on top of Oxigraph.

That's a great point. My goal in the future is to move `Store` and `Dataset` closer. The challenge with a common trait is that `Store` operations can fail because of I/O operations whereas `Dataset` operations are not supposed to fail. What we can do is make the trait operations fallible, with the implementation on `Dataset` returning the [`std::convert::Infaillible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) error.

On more concrete terms:
1. It would be great to have a `GraphView` on top of `Store`. This is not a priority for me but I would love to review and merge a pull request adding it.
2. I plan to decouple the SPARQL evaluator from the `Store` implementation, allowing to execute SPARQL query on top of `Dataset`. This would be a great occasion to improve/optimize `Dataset` internals. I will have to introduce some traits for this work that would be implemented by both `Store` and `Dataset`. This won't be done in the short term but I hope to get it done in a year or two (by free time is quite limited at the moment).

**Comment by Natureshadow** [comment]
@Tpt Looking at the base issue from a different perspective… What stops me from just using Oxigraph with the in-memory RocksDB? Has it any real disadvantages or advantages?

**Comment by Tpt** [comment]
> @Tpt Looking at the base issue from a different perspective… What stops me from just using Oxigraph with the in-memory RocksDB? Has it any real disadvantages or advantages?

Nothing really. It is slightly slower than a "in-memory" native storage and still requires access to a file system for some dumb reason but it works perfectly well. I am using it in production in a few places without trubles.

**Comment by Natureshadow** [comment]
> still requires access to a file system for some dumb reason

That implies, not usable in wasm targets, right?

**Comment by Tpt** [comment]
> That implies, not usable in wasm targets, right?

No, it is not usable sadly. However, I wrote a quite bad polyfill for this case so Oxigraph still compiles and is usable with wasm. I heop to take some time at some point to improve it and drop RocksDB usage completely for in-memory use cases
